### PR TITLE
Editor mode json schema validation

### PR DESCRIPTION
**Pull request by raphaelsolarski**

### CODE REVIEW

File: utils/components-utils/src/main/scala/pl/touk/nussknacker/engine/util/sinkvalue/SinkValueData.scala [hunk 1]
```diff
@@ -31,13 +38,40 @@ object SinkValueData {
...
+    }
+  }
 
-  case class SinkRecordParameter(fields: ListMap[FieldName, SinkValueParameter]) extends SinkValueParameter
-}
+  case class SinkRecordParameter(fields: ListMap[FieldName, SinkValueParameter]) extends SinkValueParameter {
+    override def validateParams(actualResultTypes: List[(String, BaseDefinedParameter)], path: List[String])(implicit nodeId: NodeId): ValidatedNel[ProcessCompilationError, Unit] = {
+      fields.map { case (fieldName, sinkValueParameter) =>
+        val fieldPath = path :+ fieldName
+        val fieldsMatchingPath = actualResultTypes.filter(k => k._1.startsWith(fieldPath.mkString(".")))
```
Comment (lciolecki) [inline utils/components-utils/src/main/scala/pl/touk/nussknacker/engine/util/sinkvalue/SinkValueData.scala:63]: I don't like `k._1` etc. IMO better is do: `filter{ case (realName,_ ) => }`

File: utils/schemed-kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/schemedkafka/schemaregistry/universal/ParsedSchemaSupport.scala [hunk 2]
```diff
@@ -96,16 +101,20 @@ object JsonSchemaSupport extends ParsedSchemaSupport[OpenAPIJsonSchema] {
...
 
-  override def extractSinkValueParameter(schema: ParsedSchema)(implicit nodeId: NodeId): ValidatedNel[ProcessCompilationError, SinkValueParameter] =
-    JsonSinkValueParameter(schema.cast().rawSchema(), defaultParamName = SinkValueParamName)
+  override def extractSinkValueParameter(schema: ParsedSchema, rawMode: Boolean, validationMode: ValidationMode, rawParameter: Parameter)(implicit nodeId: NodeId): ValidatedNel[ProcessCompilationError, SinkValueParameter] = {
+    if (rawMode) {
+      Validated.Valid(
+        SinkSingleValueParameter(rawParameter, new JsonSchemaOutputValidator(validationMode, schema.cast().rawSchema(), schema.cast().rawSchema()))
+      )
+    } else {
+      JsonSinkValueParameter(schema.cast().rawSchema(), defaultParamName = SinkValueParamName, ValidationMode.lax)
```
Comment (arkadius) [inline utils/schemed-kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/schemedkafka/schemaregistry/universal/ParsedSchemaSupport.scala:110]: Please add the comment why lax is hardcoded here. Especially that validationMode is available here

File: utils/schemed-kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/schemedkafka/schemaregistry/universal/UniversalSchemaSupport.scala [hunk 3]
```diff
@@ -34,10 +35,9 @@ trait UniversalSchemaSupport {
   def payloadDeserializer: UniversalSchemaPayloadDeserializer
   def serializer(schemaOpt: Option[ParsedSchema], c: SchemaRegistryClient, isKey: Boolean): Serializer[Any]
   def typeDefinition(schema: ParsedSchema): TypingResult
-  def extractSinkValueParameter(schema: ParsedSchema)(implicit nodeId: NodeId): ValidatedNel[ProcessCompilationError, SinkValueParameter]
+  def extractSinkValueParameter(schema: ParsedSchema, rawMode: Boolean, validationMode: ValidationMode, rawParameter: Parameter)(implicit nodeId: NodeId): ValidatedNel[ProcessCompilationError, SinkValueParameter]
   def sinkValueEncoder(schema: ParsedSchema, mode: ValidationMode): Any => AnyRef
-  def validateRawOutput(schema: ParsedSchema, t: TypingResult, mode: ValidationMode): ValidatedNel[OutputValidatorError, Unit]
```
Comment (arkadius) [inline utils/schemed-kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/schemedkafka/schemaregistry/universal/UniversalSchemaSupport.scala:39]: Nice, now api looks more concise +1

File: utils/schemed-kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/schemedkafka/sink/AvroSinkValueParameter.scala [hunk 4]
```diff
@@ -79,6 +79,7 @@ object AvroSinkSingleValueParameter {
       isLazyParameter = true,
       defaultValue = defaultValue.map(_.expression)
     )
-    SinkSingleValueParameter(parameter)
+    //for now we don't SchemaOutputValidator for avro in editor mode
```
Comment (arkadius) [inline utils/schemed-kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/schemedkafka/sink/AvroSinkValueParameter.scala:82]: maybe let's add TODO that we want to add this in some future?

File: utils/components-utils/src/main/scala/pl/touk/nussknacker/engine/util/sinkvalue/SinkValueData.scala [hunk 5]
```diff
@@ -31,13 +38,40 @@ object SinkValueData {
...
+    }
+  }
 
-  case class SinkRecordParameter(fields: ListMap[FieldName, SinkValueParameter]) extends SinkValueParameter
-}
+  case class SinkRecordParameter(fields: ListMap[FieldName, SinkValueParameter]) extends SinkValueParameter {
+    override def validateParams(actualResultTypes: List[(String, BaseDefinedParameter)], path: List[String])(implicit nodeId: NodeId): ValidatedNel[ProcessCompilationError, Unit] = {
+      fields.map { case (fieldName, sinkValueParameter) =>
+        val fieldPath = path :+ fieldName
+        val fieldsMatchingPath = actualResultTypes.filter { case (name, _) => name.startsWith(fieldPath.mkString(".")) }
```
Comment (arkadius) [inline utils/components-utils/src/main/scala/pl/touk/nussknacker/engine/util/sinkvalue/SinkValueData.scala:63]: IMO that `startWith` approach won't work correctly in many cases. Consider situation when you have schema like:
```
foo: String
fooBar: Integer
```
You can't be sure if `foo` will be validated against `fooBar` schema. WDYT about exact checking parameter name. I did such a change in #4013 and it looks like all tests passed. Also thanks to this approach, you don't need to rebuild "path" (one parameter less in `validateParams` method).

Comment (raphaelsolarski) [reply utils/components-utils/src/main/scala/pl/touk/nussknacker/engine/util/sinkvalue/SinkValueData.scala]: thansk, merged

File: utils/components-utils/src/main/scala/pl/touk/nussknacker/engine/util/sinkvalue/SinkValueData.scala [hunk 6]
```diff
@@ -30,14 +36,42 @@ object SinkValueData {
...
-    def toParameters: List[Parameter] = this match {
-      case SinkSingleValueParameter(value) => value :: Nil
-      case SinkRecordParameter(fields) => fields.values.toList.flatMap(_.toParameters)
+    def toParameters: List[Parameter] = flatten.map(_.value)
+
+    def flatten: List[SinkSingleValueParameter] = this match {
+      case single: SinkSingleValueParameter => single :: Nil
+      case SinkRecordParameter(fields) => fields.values.toList.flatMap(_.flatten)
     }
+    def validateParams(resultType: List[(String, BaseDefinedParameter)])(implicit nodeId: NodeId): ValidatedNel[ProcessCompilationError, Unit]
```
Comment (arkadius) [inline utils/components-utils/src/main/scala/pl/touk/nussknacker/engine/util/sinkvalue/SinkValueData.scala:45]: `List` -> `Map` ?

Comment (raphaelsolarski) [reply utils/components-utils/src/main/scala/pl/touk/nussknacker/engine/util/sinkvalue/SinkValueData.scala]: I also added type aliases to differentiate RecordFieldName and ParameterName

**Review by arkadius — APPROVED** [review]
Good job

**Pull request merged**